syntax = "proto3";

package common;

service TaskService {
    rpc CreateTask (CreateTaskRequest) returns (CreateTaskResponse);
    rpc UpdateTask (UpdateTaskRequest) returns (EmptyMessage);
    rpc DeleteTask (DeleteTaskRequest) returns (EmptyMessage);
    rpc GetTask (GetTaskRequest) returns (GetTaskResponse);
    rpc ListTasks (ListTasksRequest) returns (ListTasksResponse);

    rpc SendLike (SendLikeOrViewRequest) returns (EmptyMessage);
    rpc SendView (SendLikeOrViewRequest) returns (EmptyMessage);
    rpc GetLikesAndViews (GetLikesAndViewsRequest) returns (GetLikesAndViewsResponse);
    rpc GetTop5Posts (GetTop5PostsRequest) returns (GetTop5PostsResponse);
    rpc GetTop3Users (EmptyMessage) returns (GetTop3UsersResponse);
}

service StatService {
    rpc Healthcheck (HealthcheckRequest) returns (HealthcheckResponse);
}

message EmptyMessage {
}

message CreateTaskRequest {
    int64 author_id = 1;
    string text = 2;
}

message CreateTaskResponse {
    int64 task_id = 1;
}

message UpdateTaskRequest {
    int64 user_id = 1;
    int64 task_id = 2;
    string new_text = 3;
}

message DeleteTaskRequest {
    int64 user_id = 1;
    int64 task_id = 2;
}

message GetTaskRequest {
    int64 task_id = 1;
}

message GetTaskResponse {
    int64 task_id = 1;
    int64 author_id = 2;
    string text = 3;
}

message ListTasksRequest {
    int64 user_id = 1;
    int64 offset = 2;
    int64 limit = 3;
}

message ListTasksResponse {
    repeated Task tasks = 1;
}

message Task {
    int64 task_id = 1;
    int64 author_id = 2;
    string text = 3;
}

message SendLikeOrViewRequest {
    int64 task_id = 1;
    int64 liker_id = 2;
}

message HealthcheckRequest {
    int64 a = 1;
}

message HealthcheckResponse {
    int64 aa = 1;
}

message GetLikesAndViewsRequest {
    int64 task_id = 1;
}

message GetLikesAndViewsResponse {
    int64 task_id = 1;
    int64 likes_count = 2;
    int64 views_count = 3;
}

message GetTop5PostsRequest {
    bool sort_by_likes = 1;
}

message GetTop5PostsResponseOne {
    int64 task_id = 1;
    string author_id = 2;
    int64 likes_count = 3;
    int64 views_count = 4;
}

message GetTop5PostsResponse {
    repeated GetTop5PostsResponseOne posts = 1;
}

message GetTop3UsersResponseOne {
    string author_id = 1;
    int64 likes_count = 2;
}

message GetTop3UsersResponse {
    repeated GetTop3UsersResponseOne users = 1;
}
